File: lib/IntcodeComputer.cpp
#include "./IntcodeComputer.hpp"

IntcodeComputer::IntcodeComputer(const char path[]) {
  std::ifstream file{ path };
  if (!file) {
    std::cout << "No file" << std::endl;
  }

  std::string code{};

  while (file.good()) {
    std::getline(file, code, ',');
    intcodes.push_back(std::stoi(code));
  }

  pointer = 0;
  file.close();
  ready = true;
}

IntcodeComputer::IntcodeComputer(const char path[], std::vector<std::pair<size_t, int>> replacements) {
  *this = IntcodeComputer{ path };
  for (const auto &replacement : replacements) {
    if (!updateIntcode(replacement.first, replacement.second)) {
      std::cout << "Couldn't update intcodes with replacements" << std::endl;
      ready = false;
      break;
    }
  }
}

void IntcodeComputer::printCodes() const {
  for (size_t i{ 0 }; i < intcodes.size(); i++) {
    if (i == pointer) std::cout << ">";
    std::cout << intcodes.at(i);
    if (i == pointer) std::cout << "<";
    std::cout << " ";
  }

  std::cout << std::endl;
}

void IntcodeComputer::run() {
  if (!ready) {
    std::cout << "The computer is not ready to run" << std::endl;
    return;
  }

  while(true) {
    if (debugMode) std::cout << "Pointer position: " << pointer << std::endl;
    int verb = intcodes.at(pointer++);
    int operand1 = intcodes.at(pointer++);
    int operand2 = intcodes.at(pointer++);
    int operand3 = intcodes.at(pointer++);
    if (debugMode) std::cout << "opcode: " << verb << std::endl;
    if (debugMode) std::cout << "operand1: " << operand1 << std::endl;
    if (debugMode) std::cout << "operand2: " << operand2 << std::endl;
    if (debugMode) std::cout << "operand3: " << operand3 << std::endl;

    if (verb == 99) {
      break;
    }

    std::function<int(int, int)> op;

    if (verb == 1) {
      op = std::plus<int>();
    } else if (verb == 2) {
      op = std::multiplies<int>();
    } else {
      printDebug(verb, operand1, operand2, operand3);
      ready = false;
      break;
    }

    int result{};
    try {
      result = op(intcodes.at(operand1), intcodes.at(operand2));
    } catch (std::bad_function_call err) {
      printDebug(verb, operand1, operand2, operand3);
    }

    if (!updateIntcode(operand3, result)) {
      printDebug(verb, operand1, operand2, operand3);
      ready = false;
      break;
    }
  }
}

bool IntcodeComputer::updateIntcode(size_t index, int value) {
  if (index >= 0 && index < intcodes.size()) {
    intcodes[index] = value;
    return true;
  }

  return false;
}

void IntcodeComputer::printDebug(int opcode, int operand1, int operand2, int operand3) {
  std::cout << "Encountered an error:" << std::endl;
  std::cout << "Pointer position: " << pointer << std::endl;
  std::cout << "opcode: " << opcode << std::endl;
  std::cout << "operand1: " << operand1 << std::endl;
  std::cout << "operand2: " << operand2 << std::endl;
  std::cout << "operand3: " << operand3 << std::endl;
  printCodes();
}

int IntcodeComputer::getIntcode(size_t index) {
  return intcodes.at(index);
}

File: lib/IntcodeComputer.exs
defmodule IntcodeComputer do
  def start(a, replacements \\ %{})

  def start(path, replacements) when is_binary(path) do
    read_file(path)
    |> start(replacements)
  end

  def start(intcodes, replacements) when is_map(intcodes) do
    intcodes
    |> Map.merge(replacements)
    |> run()
  end

  def read_file(path) do
    File.read!(path)
    |> String.split(",")
    |> Stream.map(&String.to_integer/1)
    |> Stream.with_index()
    |> Stream.map(fn { code, index } -> { index, code } end)
    |> Map.new()
  end

  defp run(intcodes, pointer \\ 0) do
    case Map.fetch!(intcodes, pointer) do
      1 ->
        add(intcodes, pointer)
      2 ->
        mult(intcodes, pointer)
      99 ->
        { :ok, Map.fetch!(intcodes, 0) }
      _ ->
        { :error, :bad_opcode }
    end
  end

  defp add(intcodes, pointer) do
    do_op(intcodes, pointer, &Kernel.+/2, 4)
  end

  defp mult(intcodes, pointer) do
    do_op(intcodes, pointer, &Kernel.*/2, 4)
  end

  defp do_op(intcodes, pointer, op, 4) do
    operand1 = pointer + 1
    operand2 = pointer + 2
    operand3 = pointer + 3
    %{ ^operand1 => param1, ^operand2 => param2, ^operand3 => output } = intcodes
    Map.put(intcodes, output, op.(Map.fetch!(intcodes, param1), Map.fetch!(intcodes, param2)))
    |> run(pointer + 4)
  end
end
